src/integrated/land.jl

```diff
@@ -0,0 +1,669 @@
+        soil_args::NamedTuple = (;),
+        canopy_component_types::NamedTuple = (;),
+        canopy_component_args::NamedTuple = (;),
+        canopy_model_args::NamedTuple = (;),
+        snow_model_type::Type{SnM},
+        snow_args::NamedTuple = (;),
+        ) where {
+            FT,
+            SM <: Soil.EnergyHydrology{FT},
+            MM <: Soil.Biogeochemistry.SoilCO2Model{FT},
+            SnM <: Snow.SnowModel{FT}
+            }
+
+A constructor for the `LandModel`, which takes in the concrete model
+type and required arguments for each component, constructs those models,
+and constructs the `LandModel` from them.
+
+Each component model is constructed with everything it needs to be stepped
+forward in time, including boundary conditions, source terms, and interaction
+terms.
+"""
+function LandModel{FT}(;
+    soilco2_type::Type{MM},
+    soilco2_args::NamedTuple = (;),
+    land_args::NamedTuple = (;),
+    soil_model_type::Type{SM},
+    soil_args::NamedTuple = (;),
+    canopy_component_types::NamedTuple = (;),
+    canopy_component_args::NamedTuple = (;),
+    canopy_model_args::NamedTuple = (;),
+    snow_model_type::Type{SnM},
+    snow_args::NamedTuple = (;),
+) where {
+    FT,
+    SM <: Soil.EnergyHydrology{FT},
+    MM <: Soil.Biogeochemistry.SoilCO2Model{FT},
+    SnM <: Snow.SnowModel,
+}
+
+    (; atmos, radiation, soil_organic_carbon) = land_args
+    # These should always be set by the constructor.
+    prognostic_land_components = (:canopy, :snow, :soil, :soilco2)
+
+
+    snow = snow_model_type(;
+        boundary_conditions = Snow.AtmosDrivenSnowBC(
+            atmos,
+            radiation,
+            prognostic_land_components,
+        ),
+        snow_args...,
+    )
+    # soil
+    sources = (RootExtraction{FT}(), Soil.PhaseChange{FT}())
+    if :runoff ∈ propertynames(land_args)
+        top_bc = ClimaLand.AtmosDrivenFluxBC(
+            atmos,
+            radiation,
+            land_args.runoff,
+            prognostic_land_components,
+        )
+    else #no runoff model
+        top_bc = ClimaLand.AtmosDrivenFluxBC(
+            atmos,
+            radiation,
+            ClimaLand.Soil.Runoff.NoRunoff(),
+            prognostic_land_components,
+        )
+    end
+    zero_flux = Soil.HeatFluxBC((p, t) -> 0.0)
+    boundary_conditions = (;
+        top = top_bc,
+        bottom = Soil.WaterHeatBC(;
+            water = Soil.FreeDrainage(),
+            heat = zero_flux,
+        ),
+    )
+    soil = soil_model_type(;
+        boundary_conditions = boundary_conditions,
+        sources = sources,
+        soil_args...,
+    )
+
+    transpiration = Canopy.PlantHydraulics.DiagnosticTranspiration{FT}()
+    ground_conditions =
+        PrognosticGroundConditions{FT, typeof(soil.parameters.PAR_albedo)}(
+            snow.parameters.α_snow,
+            soil.parameters.PAR_albedo,
+            soil.parameters.NIR_albedo,
+        )
+    if :energy in propertynames(canopy_component_args)
```

Comment:
```suggestion
    if :energy in propertynames(canopy_component_args)
        energy = canopy_component_types.energy(
                canopy_component_args.energy.parameters,
            )
    else
        energy = PrescribedCanopyTempModel{FT}()
    end
```

Then we can construct the `CanopyModel` the same way for both cases, setting `energy = energy` for the energy model. This will minimize the code in branches and make the code more readable

Reply:
we could also do this for runoff when we construct the `ClimaLand.AtmosDrivenFluxBC` if you like this design pattern

Reply:
this is much cleaner! Ill adopt it.

Reply:
also for runoff

Reply:
I don't think this has been changed

Reply:
ah, I only changed it in soil_canopy_model. thanks for catching this!

Reply:
oh wait no, it is changed here too. right?

Reply:
oh yes it is! Sorry about that